### PR TITLE
Optimize unpaginated collection_versions endpoint query

### PR DESCRIPTION
**Pull request by gerrod3**
[noissue]

This should reduce the amount of database queries 5x. So now the amount of queries for this endpoint is n + 4 where n is the number of collectionversions.  I'm not sure if I can remove the final n queries to make this constant since they are for the tags which are a many-to-many relation.

### CODE REVIEW

File: pulp_ansible/app/galaxy/v3/views.py
Comment (fao89) [inline pulp_ansible/app/galaxy/v3/views.py:447]: we need to sort by version

File: pulp_ansible/app/galaxy/v3/views.py
Comment (fao89) [inline pulp_ansible/app/galaxy/v3/views.py:474]: I don't know much about django ORM, but this method seems to be rebuilding some django method

File: pulp_ansible/app/galaxy/v3/views.py
Comment (gerrod3) [reply pulp_ansible/app/galaxy/v3/views.py:447]: Why though? It's just a json dict that we are returning, does galaxy expect the dictionary to be ordered?

Comment (fao89) [reply pulp_ansible/app/galaxy/v3/views.py:447]: I don't know exactly the `ansible-galaxy` CLI code, but I think it expects the versions to be ordered:
https://galaxy.ansible.com/api/v2/collections/pulp/pulp_installer/versions/

Comment (gerrod3) [reply pulp_ansible/app/galaxy/v3/views.py:447]: But this endpoint is only for syncing, correct? They would still use the other endpoints that sort by version for installing collections.

Comment (fao89) [reply pulp_ansible/app/galaxy/v3/views.py:447]: Actually I checked the code and they just loop through versions 
https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/api.py#L767-L847

Comment (fao89) [reply pulp_ansible/app/galaxy/v3/views.py:447]: you are right! `UnpaginatedCollectionVersionViewSet` is only used on sync

Comment (bmbouter) [reply pulp_ansible/app/galaxy/v3/views.py:447]: Dropping the ordering requirement probably would make it go faster.

Comment (bmbouter) [reply pulp_ansible/app/galaxy/v3/views.py:447]: and I agree it's only used on sync

File: pulp_ansible/app/galaxy/v3/views.py
Comment (bmbouter) [reply pulp_ansible/app/galaxy/v3/views.py:474]: I typically use cprofile to get python code timings and visualize them. (lmk if you want any help).

Also getting the SQL EXPLAIN statements for the SQL produced is then more easily analyzed to show 'oh look these queries are better/faster'.

File: pulp_ansible/app/galaxy/v3/views.py
Comment (dralley) [inline pulp_ansible/app/galaxy/v3/views.py:488]: A list comprehension would probably be clearer in this case.

**Comment by gerrod3** [comment]
@bmbouter Here is some comparison data. I synced pulp_installer which has 33 collection versions.

**Without optimization:**
Screenshot from 2021-05-12 17-35-36

**With optimization:**
Screenshot from 2021-05-12 17-36-27

The old way does 4 + 5 * n database queries (4 + 5 * 33 = 169) while the new optimized way does 4 + n (4 + 33 = 37).  The new way alone is not enough to fix the timeout issues with large collection versions count (2000+), but this optimization with streaming endpoint #562 does fix the timeout issue for repositories up to 7000 collection versions. (as far as I tested)

These two changes only move the goal post down further and can't be a permanent solution. Do you think it's still worth to try them?

**Comment by bmbouter** [comment]
Thank you for posting this and doing the analysis. I think I'm reading that the streaming option and this query optimization PR together resolved the issue as high as you've tested 7K collections. That's great. I'm suspecting of the two, the streaming resolution will resolve it on its own because even if the queries take longer, we can just make the batches smaller until it also works up to 7K.

If that's the case, that's what I'd recommend (keep the streaming resolution and not merge this query change). My thinking is that this mostly makes the ORM faster and it's not worth carrying that if we don't need to.

@gerrod3 would you be willing to test the "largest collection count you tested" (iirc 7K?) against just the streaming PR? And if it doesn't work maybe adjust the batch sizes to be smaller? Or is this not a great idea? wdyt?

**Comment by gerrod3** [comment]
> @gerrod3 would you be willing to test the "largest collection count you tested" (iirc 7K?) against just the streaming PR? And if it doesn't work maybe adjust the batch sizes to be smaller? Or is this not a great idea? wdyt?

I'm not sure adjusting the batch_sizes will make the streaming PR pass on its own, but maybe it will.  I can adjust this PR to be less crazy, but still have some optimizations.  Like I said in our meeting, the 5 queries per collection version include 2 duplicates and 1 that could be removed with select_related (the current select_related is actually doing nothing on the query). Less queries will make the streaming endpoint faster.

**Comment by bmbouter** [comment]
Also we could accept as-is and move on to other work? Overall I don't think this code is going to be here that long because I believe we need to switch pulp_ansible to use publications and have this be pre-computed instead. If we are going to keep this code around then I think we do need to be careful, but if not, not really.

@gerrod3 @daviddavis @fao89 wdyt should be done? Also are there concerns about converting pulp_ansible to a publication-based plugin soon? Any feedback is welcome.

**Comment by gerrod3** [comment]
@bmbouter @fao89 @daviddavis I just simplified the optimization so now the amount of requests is 4 + 2n.  I think this is much more manageable and still provides much better performance. I think we should merge this variation.

**Comment by bmbouter** [comment]
> @bmbouter @fao89 @daviddavis I just simplified the optimization so now the amount of requests is 4 + 2n. I think this is much more manageable and still provides much better performance. I think we should merge this variation.

This looks good to me. Is the stremaing response one already merged?

**Comment by fao89** [comment]
This PR looked good to me before, I didn't approve so far because I was hoping someone would jump in and say: "here, use this django_orm_magic_method instead"

**Review by fao89 — COMMENTED** [review]
But we need a changelog entry for this!
Other than that, consider this PR approved!

**Comment by gerrod3** [comment]
> This looks good to me. Is the stremaing response one already merged?

The streaming response #562 is dependent on this PR.

@fao89 Added changed log.

**Review by fao89 — APPROVED** [review]
Thank you!

**Comment by bmbouter** [comment]
Agreed. Thank you!

**Comment by bmbouter** [comment]
@gerrod3 ready for merging?

**Pull request merged**
